### PR TITLE
Convert the LocalExecutor to run tasks using new Task SDK supervisor code

### PR DESCRIPTION
**Pull request by ashb**
This also lays the groundwork for a more general purpose "workload" execution
system, make a single interface for executors to run tasks and callbacks.

Also in this PR we set up the supervise function to send Task logs to a file,
and handle the task log template rendering in the scheduler before queueing
the workload.

Additionally we don't pass the activity directly to `supervise()` but instead
the properties/fields of it to reduce the coupling between SDK and Executor.
(More separation will appear in PRs over the next few weeks.)

The big change of note here is that rather than sending an airflow command
line to execute (`["airflow", "tasks", "run", ...]`) and going back in via the
CLI parser we go directly to a special purpose function. Much simpler.

It doesn't remove any of the old behaviour (CeleryExecutor still uses
LocalTaskJob via the CLI parser etc.), nor does anything currently send
callback requests via this new workload mechanism.

The `airflow.executors.workloads` module currently needs to be shared between
the Scheduler (or more specifically the Executor) and the "worker" side of
things. In the future these will be separate python dists and this module will
need to live somewhere else.

Right now we check the if `executor.queue_workload` is different from the
BaseExecutor version (which just raises an error right now) to see which
executors support this new version. That check will be removed as soon as all
the in-tree executors have been migrated.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/executors/local_executor.py [hunk 1]
```diff
@@ -83,88 +75,47 @@ def _run_worker(
...
-        logging.shutdown()
-        os._exit(ret)
+    setproctitle(f"airflow worker -- LocalExecutor: {workload.ti.id}")
+    # This will return the exit code of the task process, but we don't care about that, just if the
+    # _supervisor_ had an error reporting the state back (which will result in an exception.)
+    supervise(
+        ti=workload.ti,
+        dag_path=workload.dag_path,
+        token=workload.token,
+        server="http://localhost:9091/execution/",
```
Comment (kaxil) [inline airflow/executors/local_executor.py]: This _should_ come from a config/env var?

Comment (ashb) [reply airflow/executors/local_executor.py]: In the case of LocalExecutor specifically I'm not sure -- it should always be localhost shouldn it it?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 2]
```diff
@@ -297,6 +298,10 @@ def start(
         if pid == 0:
             # Parent ends of the sockets are closed by the OS as they are set as non-inheritable
 
+            del path
+            del client
+            del ti
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py:304]: Why delete this? want to remove accidental use?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 3]
```diff
@@ -369,7 +378,7 @@ def _close_unused_sockets(*sockets):
 
     def _send_startup_message(self, ti: TaskInstance, path: str | os.PathLike[str], child_comms: socket):
         """Send startup message to the subprocess."""
-        msg = StartupDetails(
+        msg = StartupDetails.model_construct(
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py:383]: Why this change? for skipping validation?

File: task_sdk/src/airflow/sdk/execution_time/task_runner.py [hunk 4]
```diff
@@ -37,9 +37,9 @@
 
 
 class RuntimeTaskInstance(TaskInstance):
-    model_config = ConfigDict(arbitrary_types_allowed=True)
+    model_config = ConfigDict(arbitrary_types_allowed=True, defer_build=True)
 
-    task: BaseOperator
+    task: Annotated[BaseOperator, SkipValidation]
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/task_runner.py]: Do we need to use both `defer_build=True` & `SkipValidation`?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 2 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:304]: Likely not required, but just to make 100% sure they get gc'd and don't leak around longer than they are needed

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 3 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:383]: Yeah.

File: task_sdk/src/airflow/sdk/execution_time/task_runner.py (hunk 4 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/task_runner.py]: I think so, defer_build is about building the type schema etc and happened even with SkipValidation.

I think without this the error was something imported from BaseOperator failed to evaluate a type annotation.

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/task_runner.py]: Though maybe we don't need this and the `model_construct` call

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/task_runner.py]: Yeah actually, both of these can go, `.model_construct()` alone is enough (as it skips all validation).

For the record, the error without it is:

> pydantic.errors.PydanticUserError: `RuntimeTaskInstance` is not fully defined; you should define `TaskGroup`, then call `RuntimeTaskInstance.model_rebuild()`.

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/task_runner.py]: Removed these now.

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 3 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:383]: Also: by using this it doesn't need BaseOperator _and everything it imports_ to be loaded and defined for pydantic to build the schema/validation logic internally.

File: airflow/executors/local_executor.py (hunk 1 above)
Comment (kaxil) [reply airflow/executors/local_executor.py]: Probably, but `port` could change if there are other things on the box?

**Review by kaxil — APPROVED** [review]
Minor comments/questions but directionally lgtm

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 2 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:304]: Worth adding a comment?

File: airflow/executors/local_executor.py (hunk 1 above)
Comment (ashb) [reply airflow/executors/local_executor.py]: Yeah fair. Where do you think we should pull this from right now?

Comment (kaxil) [reply airflow/executors/local_executor.py]: airflow.cfg is fine for now? Worst case, an env var for now -- to be changed later?

Comment (jscheffl) [reply airflow/executors/local_executor.py]: Is `conf.getint()` not suiteable for this?

File: airflow/executors/workloads.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    task_id: str
+    dag_id: str
+    run_id: str
+    try_number: int
+    map_index: int | None = None
+
+    # TODO: Task-SDK: see if we can replace TIKey with this class entirely?
+    @property
+    def key(self) -> TaskInstanceKey:
```
Comment (jscheffl) [inline airflow/executors/workloads.py]: I do not fully get this. So yes, atm the `TaskInstance` class is very base and has the same content like the former `TaskInstanceKey` class - but I do not understand why it is not needing more? And why if not more is needed do we not just name this the `TaskInstanceKey`? I mean it carries only PK fields, no value as payload.

File: airflow/executors/workloads.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        from airflow.models.taskinstancekey import TaskInstanceKey
+
+        return TaskInstanceKey(
+            dag_id=self.dag_id,
+            task_id=self.task_id,
+            run_id=self.run_id,
+            try_number=self.try_number,
+            map_index=-1 if self.map_index is None else self.map_index,
```
Comment (jscheffl) [inline airflow/executors/workloads.py:63]: Can be a bit easier :-D
```suggestion
            map_index=self.map_index or -1,
```

File: airflow/executors/workloads.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+class ExecuteTask(BaseActivity):
+    """Execute the given Task."""
+
+    ti: TaskInstance
+    """The TaskInstance to execute"""
+    dag_path: os.PathLike[str]
+    """The filepath where the DAG can be found (likely prefixed with `DAG_FOLDER/`)"""
+
+    log_filename_suffix: str | None
+    """The rendered log filename template the task logs should be written to"""
```
Comment (jscheffl) [inline airflow/executors/workloads.py]: Why is this a "suffix"? Because it is a relative path to the log root? `suffix` sounds more like `.log` as file extension. Why not...
```suggestion
    log_path: Path | None
    """The rendered relative log filename and path the task logs should be written to"""
```

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 8]
```diff
@@ -630,25 +638,57 @@ def forward_to_log(target_log: FilteringBoundLogger, level: int) -> Generator[No
             target_log.log(level, msg)
 
 
-def supervise(activity: ExecuteTaskActivity, server: str | None = None, dry_run: bool = False) -> int:
+def supervise(
+    *,
+    ti: Any,
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Why `Any`?

**Review by jscheffl — APPROVED** [review]
I do not fully understand the code but maybe I need a night of sleep. Just some comments not blocking.

I was scratching my head regarding the general supervisor approach - Do we have platform limitations by this? I assume no, all major/general operating systems have the socket mechanisms, also Windows, correct? Do we see limitations for running outside Linux?

File: airflow/executors/workloads.py (hunk 5 above)
Comment (ashb) [reply airflow/executors/workloads.py]: This is a compat shim to not need to change too much -- this is called `key` because that is the method that exist on `airflow.models.taskinstance.TaskInstance`.

Longer term, the uuid in `id` should be all we need and we can remove the concept of TI key.

File: airflow/executors/workloads.py (hunk 6 above)
Comment (ashb) [reply airflow/executors/workloads.py:63]: Using `None` was intentional. `-1` was only used because map_index forms part of the (old) PK of the `task_instance` table, and you can't have a nullable column in the PK. None is a better representation of "not mapped" than -1 is.

(Also `self.map_index or -1` would turn `0` into `-1` which we don't want)

File: airflow/executors/workloads.py (hunk 7 above)
Comment (ashb) [reply airflow/executors/workloads.py]: Yeah, good call.

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 8 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Typing laziness, and trying to not create deps. But I think the `Any` came before we had a (simplified) generated form of TI in the API. I'll swap it over to that.

**Comment by ashb** [comment]
Yes, windows supports the socketpair etc (it's why I chose that API). This supervisor approach is what we do today with LocalTaskJob and the StandardRunner etc, this is just a simplified re-implementation of it.

To fully support windows we will need to have something other than `os.fork` based approach, the rest should be possible. Windows supports keeping open "inheritable" sockets open to launched processes, so it'll be slower because of launching a whole new python interpreter, but again, this is the current behaviour on Windows of StandardTask runner.

File: airflow/executors/workloads.py (hunk 5 above)
Comment (ashb) [reply airflow/executors/workloads.py]: I'll update this comment with `Can we replace TIKey with just the uuid?`

File: airflow/executors/local_executor.py (hunk 1 above)
Comment (ashb) [reply airflow/executors/local_executor.py]: Either conf.getint for just the port, or maybe we just have a config option of "execution_api_server_url"

Comment (ashb) [reply airflow/executors/local_executor.py]:
```python
        server=conf.get("workers", "execution_api_server_url", fallback="http://localhost:9091/execution/"),
``` 

is what I'm going with.

File: airflow/executors/workloads.py (hunk 6 above)
Comment (jscheffl) [reply airflow/executors/workloads.py:63]: Oh, forgot about the `0`case :-O

File: airflow/executors/local_executor.py (hunk 1 above)
Comment (kaxil) [reply airflow/executors/local_executor.py]: Yeah that sounds good
